File: resources/patterns/ai/prompt-chaining-pattern/prompt_chain.py
"""Prompt Chaining Pattern - Python implementation.

Chain multiple LLM calls where each step's output feeds the next step's input.
Break complex tasks into smaller, verifiable prompts for better results.
"""

from dataclasses import dataclass
from typing import Callable, List, Optional


@dataclass
class ChainStep:
    name: str
    prompt_template: str
    model: str = "gpt-4o"
    temperature: float = 0.7
    max_retries: int = 2
    validator: Optional[Callable[[str], bool]] = None


@dataclass
class ChainResult:
    step_name: str
    input: str
    output: str
    success: bool
    retries: int


def mock_llm_call(prompt: str, model: str, temperature: float) -> str:
    """Simulate an LLM API call. Replace with your actual LLM client."""
    return f"[{model}] Processed: {prompt[:60]}..."


class PromptChain:
    def __init__(self, steps: List[ChainStep]):
        self.steps = steps

    def run(self, initial_input: str) -> List[ChainResult]:
        results: List[ChainResult] = []
        current_input = initial_input

        for step in self.steps:
            prompt = step.prompt_template.format(input=current_input)
            success = False
            output = ""
            retries = 0

            for attempt in range(step.max_retries + 1):
                output = mock_llm_call(prompt, step.model, step.temperature)
                retries = attempt

                if step.validator and not step.validator(output):
                    print(f"  Step '{step.name}' validation failed (attempt {attempt + 1})")
                    continue

                success = True
                break

            results.append(ChainResult(
                step_name=step.name,
                input=current_input,
                output=output,
                success=success,
                retries=retries,
            ))

            if not success:
                print(f"Chain stopped at step '{step.name}' after {retries + 1} attempts")
                break

            current_input = output
            print(f"Step '{step.name}' completed")

        return results


# --- Validators ---

def validate_non_empty(text: str) -> bool:
    return len(text.strip()) > 10


def validate_has_code(text: str) -> bool:
    return "```" in text or "def " in text or "function " in text


# --- Usage example ---

if __name__ == "__main__":
    chain = PromptChain([
        ChainStep(
            name="extract_requirements",
            prompt_template="Extract key requirements from this text:\n{input}\n\nList each requirement as a bullet point.",
            model="gpt-4o",
            temperature=0.2,
            validator=validate_non_empty,
        ),
        ChainStep(
            name="generate_code",
            prompt_template="Based on these requirements, generate code:\n{input}\n\nProvide working code with comments.",
            model="gpt-4o",
            temperature=0.3,
            validator=validate_has_code,
        ),
        ChainStep(
            name="review_code",
            prompt_template="Review this code for bugs and improvements:\n{input}\n\nList issues found.",
            model="gpt-4o",
            temperature=0.5,
            validator=validate_non_empty,
        ),
        ChainStep(
            name="format_report",
            prompt_template="Create a summary report from this review:\n{input}\n\nFormat as markdown with sections.",
            model="gpt-4o-mini",
            temperature=0.7,
        ),
    ])

    results = chain.run("Build a REST API endpoint that accepts JSON, validates input, and returns 201 on success")

    print(f"\nChain completed: {len(results)}/{len(chain.steps)} steps")
    for r in results:
        status = "OK" if r.success else "FAILED"
        print(f"  {r.step_name}: {status} (retries: {r.retries})")

File: resources/patterns/ai/prompt-chaining-pattern/prompt_chain.js
/**
 * Prompt Chaining Pattern - JavaScript implementation.
 *
 * Chain multiple LLM calls where each step's output feeds the next step's input.
 * Break complex tasks into smaller, verifiable prompts for better results.
 */

class ChainStep {
  constructor(name, promptTemplate, options = {}) {
    this.name = name;
    this.promptTemplate = promptTemplate;
    this.model = options.model || "gpt-4o";
    this.temperature = options.temperature ?? 0.7;
    this.maxRetries = options.maxRetries ?? 2;
    this.validator = options.validator || null;
  }
}

/**
 * Simulate an LLM API call. Replace with your actual LLM client.
 */
function mockLlmCall(prompt, model, temperature) {
  return `[${model}] Processed: ${prompt.slice(0, 60)}...`;
}

class PromptChain {
  constructor(steps) {
    this.steps = steps;
  }

  async run(initialInput) {
    const results = [];
    let currentInput = initialInput;

    for (const step of this.steps) {
      const prompt = step.promptTemplate.replace("{input}", currentInput);
      let success = false;
      let output = "";
      let retries = 0;

      for (let attempt = 0; attempt <= step.maxRetries; attempt++) {
        output = mockLlmCall(prompt, step.model, step.temperature);
        retries = attempt;

        if (step.validator && !step.validator(output)) {
          console.log(`  Step '${step.name}' validation failed (attempt ${attempt + 1})`);
          continue;
        }

        success = true;
        break;
      }

      results.push({ stepName: step.name, input: currentInput, output, success, retries });

      if (!success) {
        console.log(`Chain stopped at step '${step.name}' after ${retries + 1} attempts`);
        break;
      }

      currentInput = output;
      console.log(`Step '${step.name}' completed`);
    }

    return results;
  }
}

// --- Validators ---

function validateNonEmpty(text) {
  return text.trim().length > 10;
}

function validateHasCode(text) {
  return text.includes("```") || text.includes("function ") || text.includes("const ");
}

// --- Usage example ---

const chain = new PromptChain([
  new ChainStep("extract_requirements", "Extract key requirements from:\n{input}\n\nList as bullet points.", { temperature: 0.2, validator: validateNonEmpty }),
  new ChainStep("generate_code", "Generate code from these requirements:\n{input}\n\nProvide working code.", { temperature: 0.3, validator: validateHasCode }),
  new ChainStep("review_code", "Review this code for bugs:\n{input}\n\nList issues found.", { temperature: 0.5, validator: validateNonEmpty }),
  new ChainStep("format_report", "Create a summary report:\n{input}\n\nFormat as markdown.", { model: "gpt-4o-mini", temperature: 0.7 }),
]);

chain.run("Build a REST API endpoint that accepts JSON, validates input, and returns 201 on success").then(results => {
  console.log(`\nChain completed: ${results.length}/${chain.steps.length} steps`);
  results.forEach(r => {
    const status = r.success ? "OK" : "FAILED";
    console.log(`  ${r.stepName}: ${status} (retries: ${r.retries})`);
  });
});
